File: AccuPay/Core/Utilities/Extensions/DecimalExtensions.vb
Option Strict On

Imports System.Runtime.CompilerServices
Imports AccuPay.Utils

Namespace Global.AccuPay.Extensions
    Public Module DecimalExtensions

        <Extension()>
        Public Function RoundToString(input As Decimal, Optional decimalPlace As Integer = 2) As String

            Dim format = "#,##0."

            For i = 1 To decimalPlace
                format &= "0"
            Next

            Return AccuMath.CommercialRound(input, decimalPlace).ToString(format)

        End Function

        Public Function RoundToString(input As Decimal?, Optional decimalPlace As Integer = 2) As String

            If input Is Nothing Then input = 0

            Dim format = "#,##0."

            For i = 1 To decimalPlace
                format &= "0"
            Next

            Return AccuMath.CommercialRound(input, decimalPlace).ToString(format)

        End Function

    End Module

End Namespace

File: AccuPay.Utilities/Extensions/DecimalExtensions.cs
namespace AccuPay.Utilities.Extensions
{
    public static class DecimalExtensions
    {
        public static string RoundToString(this decimal input, int decimalPlace = 2)
        {
            string format = "#,##0.";

            // TODO: check in benchmark if this shows the right format
            for (int i = 1; i < decimalPlace; i++)
            {
                format += "0";
            }

            return AccuMath.CommercialRound(input, decimalPlace).ToString(format);
        }

        public static string RoundToString(this decimal? input, int decimalPlace = 2)
        {
            if (input == null) input = 0;

            string format = "#,##0.";

            for (int i = 1; i <= decimalPlace; i++)
            {
                format += "0";
            }

            return AccuMath.CommercialRound(input, decimalPlace).ToString(format);
        }

        public static decimal Round(this decimal input, int decimalPlace = 2)
        {
            return AccuMath.CommercialRound(input, decimalPlace);
        }
    }
}
